src/header-utilities/exponentialSearch.hpp

```diff
@@ -0,0 +1,51 @@
+class ExponentialSearch {
+  template< typename RandomAccessIterator,
+	    typename Value,
+	    typename Compare >
+  static RandomAccessIterator apply( RandomAccessIterator left,
+				     RandomAccessIterator right,
+				     RandomAccessIterator end,
+				     Value&& value,
+				     Compare&& compare ){
```

Comment:
The interface is based on [lower_bound](http://en.cppreference.com/w/cpp/algorithm/lower_bound) in the standard library algorithm header. A `Compare` type provides an overload of the parentheses operator which accepts two argument and specifies whether the left argument is less than the right argument, for some definition of less. In this sense, it defines the expected ordering of the sequence being searched.

That's a lot of mumbo jumbo, so I think an example might be helpful. Suppose I have a class `Employee`.

```cpp
struct Employee {
  std::string name;
  std::string znumber;
};
```

Given a sequence of Employees (say a `std::vector<Employee>`), I could sort the sequence based on the `name` field or the `znumber` field. If, after sorting, I want to search the sequence, I need to specify how the employees are ordered. I do this by providing a comparison function.

```cpp
std::vector<Employee> employees = {...};
auto byName = 
  []( const Employee& left, const Employee& right ){ return left.name < right.name; };

std::sort( employee.begin(), employees.end(), byName );

Employee methuselah{ 'methuselah', '000001' };
auto iterator = std::lower_bound( employee.begin(), employees.end(), methuselah, byName );
```